libs/vkd3d/va_map.c

```diff
@@ -0,0 +1,382 @@
+}
+
+static const struct vkd3d_va_block *vkd3d_va_map_find_block(const struct vkd3d_va_map *va_map, VkDeviceAddress va)
+{
+    VkDeviceAddress next_address = vkd3d_va_map_get_next_address(va);
+    const struct vkd3d_va_tree *tree = &va_map->va_tree;
+
+    while (next_address && tree)
+    {
+        tree = tree->next[next_address & VKD3D_VA_NEXT_MASK];
+        next_address >>= VKD3D_VA_NEXT_BITS;
+    }
+    
+    if (!tree)
+        return NULL;
+
+    return &tree->blocks[vkd3d_va_map_get_block_address(va)];
+}
+
+static struct vkd3d_va_block *vkd3d_va_map_get_block(struct vkd3d_va_map *va_map, VkDeviceAddress va)
+{
+    VkDeviceAddress next_address = vkd3d_va_map_get_next_address(va);
+    struct vkd3d_va_tree *tree, **tree_ptr;
+    
+    tree = &va_map->va_tree;
+
+    while (next_address)
+    {
+        tree_ptr = &tree->next[next_address & VKD3D_VA_NEXT_MASK];
+
+        if (!(tree = *tree_ptr))
+        {
+            void *orig;
+            tree = vkd3d_malloc(sizeof(*tree));
+            memset(tree, 0, sizeof(*tree));
+
+            orig = vkd3d_atomic_ptr_compare_exchange(tree_ptr, NULL, tree, vkd3d_memory_order_release, vkd3d_memory_order_relaxed);
+
+            if (orig)
+            {
+                vkd3d_free(tree);
+                tree = orig;
+            }
+        }
+
+        next_address >>= VKD3D_VA_NEXT_BITS;
+    }
+    
+    return &tree->blocks[vkd3d_va_map_get_block_address(va)];
+}
+
+static void vkd3d_va_map_cleanup_tree(struct vkd3d_va_tree *tree)
+{
+    unsigned int i;
+
+    for (i = 0; i < ARRAY_SIZE(tree->next); i++)
+    {
+        if (tree->next[i])
+        {
+            vkd3d_va_map_cleanup_tree(tree->next[i]);
+            vkd3d_free(tree->next[i]);
+        }
+    }
+}
+
+static const struct vkd3d_unique_resource *vkd3d_va_map_find_entry(struct vkd3d_va_map *va_map,
+        VkDeviceAddress va, size_t *index)
+{
+    const struct vkd3d_unique_resource *resource = NULL;
+    size_t hi = va_map->small_entries_count;
+    size_t lo = 0;
+
+    while (lo < hi)
+    {
+        const struct vkd3d_unique_resource *r;
+        size_t i = lo + (hi - lo) / 2;
+
+        r = va_map->small_entries[i];
+
+        if (va < r->va)
+            hi = i;
+        else if (va >= r->va + r->size)
+            lo = i + 1;
+        else
+        {
+            lo = hi = i;
+            resource = r;
+        }
+    }
+
+    if (index)
+        *index = lo;
+
+    return resource;
+}
+
+void vkd3d_va_map_insert(struct vkd3d_va_map *va_map, const struct vkd3d_unique_resource *resource)
+{
+    VkDeviceAddress block_va, min_va, max_va;
+    struct vkd3d_va_block *block;
+    size_t index;
+
+    if (resource->size >= VKD3D_VA_BLOCK_SIZE)
+    {
+        min_va = resource->va;
+        max_va = resource->va + resource->size;
+        block_va = min_va & ~VKD3D_VA_LO_MASK;
+
+        while (block_va < max_va)
+        {
+            block = vkd3d_va_map_get_block(va_map, block_va);
+
+            if (block_va < min_va)
```

Comment:
Don't we somehow need atomic store here at least?

Reply:
I guess every single access to the block resource pointers and va must be atomic.

Reply:
Should be fixed, needs a second pair of eyes though.

Reply:
Ye, I think relaxed is fine here since other threads need a memory barrier to be able to even observe the newly allocated VA anyways.